crowdflower_results_into_training_data.py

```diff
@@ -7,138 +7,189 @@
+
     return processed
 
 
 def set_majority_vote_answer(results_json):
+    """ Determine the correct entity which corresponds to a given FE """
+
     for k,v in results_json.iteritems():
         for fe in v.keys():
-            if fe != 'frame' and fe != 'lu' and fe != 'sentence':
-                answers_count = Counter(v[fe]['answers'])
-                majority = float(v[fe]['judgments'])/2.0
-                for answer,freq in answers_count.iteritems():
-                    if freq > majority: v[fe]['majority'] = answer
-                if not v[fe].get('majority'): print "HEADS UP! No majority answer for sentence [%s], FE [%s]" % (k, fe)
-    return results_json
+            if fe in {'frame', 'lu', 'sentence'}:
+                continue
+
+            answers_count = Counter(v[fe]['answers'])
+            majority = v[fe]['judgments'] / 2.0
+            for answer,freq in answers_count.iteritems():
+                if freq > majority:
+                    v[fe]['majority'] = answer
+
+            if not v[fe].get('majority'):
+                print "HEADS UP! No majority answer for sentence [%s], FE [%s]" % (k, fe)
+
+
+def tag_entities(results):
+    """ Creates the IOB tag for each entity found. """
+    for annotations in results.values():
+        frame = annotations['frame']
+
+        annotations['entities'] = dict()
+        for fe in annotations.keys():
+            if fe in {'frame', 'lu', 'sentence'}:
+                continue
+
+            # skip uncertain answers
+            annotation = annotations[fe].get('majority')
+            if not annotation:
+                continue
+
+            # build token label using token position, FE and frame
+            label = '%s_%s' % (fe, frame)
+            iob_tagged = [ (token, '%s-%s' % ('B' if i == 0 else 'I', label))
+                for i, token in enumerate(annotation.split())
+            ]
+            annotations['entities'][fe] = iob_tagged
+
+
+def process_sentence(sentence_id, annotations, lines):
+    """ Processes a sentence by merging tagged words, LU and FEs """
+
+    processed = list()
+    for i, (token, pos, lemma) in enumerate(lines):
+        # TODO check if LUs can be more than one token
+        tag = 'B-LU' if lemma == annotations['lu'] else 'O'
+        processed.append([
+            sentence_id, str(i), token, pos, lemma, annotations['frame'], tag
+        ])
+
+    # find the entities in the sentence and set iob tags accordingly
+    # checking for single tokens is not enough, entities have to be matched as a
+    # whole (i.e. all its tokens must appear in sequence)
+    for entity, tokens in annotations['entities'].iteritems():
+        found = False
+        i = j = 0
+        while i < len(processed):
+            if processed[i][2] == tokens[j][0]:
+                j += 1
+                if j == len(tokens):
+                    found = True
+                    break
+            else:
```

Comment:
I don't the get the reason of this `else`. Can you explain why you put it?
`j` is initialized to `0` before the loop. At runtime, if we enter the `if` block, `j` increments, otherwise nothing happens (i.e., it should still be `= 0`).
Or did I miss something?